File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        n = len(s)
        char_map = {}
        max_length = 0
        left = 0

        for right in range(n):
            current_char = s[right]
            
            if current_char in char_map and char_map[current_char]>=left:
                left = char_map[current_char]+1

            char_map[current_char] = right
            max_length = max(max_length, right-left+1)
        
        return max_length




        

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    // //maxString = new Set
    // let subString = new Set();
    // let maxLength = 0;
    // //maxlength = 0
    // //loop through string
    // for (let i=0; i<s.length; i++) {
    //     //if str[i] not in maxString
    //     // if (subString.has(s[i])) {
    //     //     //start new maxString set with str[i] 
    //     //     subString = new Set(s[i]);

    //     // } else {
    //     //     // add str[i] to the set 
    //     //     //currentLength = maxString.length
    //     //     ////maxLength = max(maxLength, currentLength)
    //     //     subString.add(s[i]);
    //     //     maxLength = Math.max(maxLength, subString.size);
    //     // }
            
    // }

    let maxLength = 0;
    let start = 0;
    let subStringMap = new Map();

    for(let i=0;  i<s.length; i++) {

        if (subStringMap.has(s[i]) && subStringMap.get(s[i])>=start) {
            start = subStringMap.get(s[i]) + 1;
        } 
        subStringMap.set(s[i],i);
        maxLength = Math.max(maxLength, (i-start+1));

    }          
    return maxLength;
};